### PR TITLE
 XeTwitterLink added

### PR DESCRIPTION
**Pull request by ikhvostenkov**
This is fix for https://github.com/yegor256/takes/issues/287. Added ```XeTwitterLink``` class and ```XeTwitterLinkTest``` test for it.

### CODE REVIEW

**Comment by ikhvostenkov** [comment]
@davvd PR is ready for review.

**Comment by davvd** [comment]
@ikhvostenkov Thanks, someone will review your pull request soon

**Comment by davvd** [comment]
@ggajos review this one, please

**Comment by davvd** [comment]
> @davvd PR is ready for review.

@ikhvostenkov I will take a look, thanks

File: src/test/java/org/takes/facets/auth/social/XeTwitterLinkTest.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ * @since 1.0
+ */
+public final class XeTwitterLinkTest {
+
+    /**
+     * XeTwitterLink can create a correct link.
+     * @throws IOException If some problem inside
+     */
+    @Test
+    public void generatesCorrectLink() throws IOException {
```
Comment (ggajos) [inline src/test/java/org/takes/facets/auth/social/XeTwitterLinkTest.java]: @ikhvostenkov we should change description to `generate` or name to `creates`

File: src/test/java/org/takes/facets/auth/social/XeTwitterLinkTest.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+                "/root/links/link[@rel='takes:twitter']"
+            )
+        );
+    }
+    /**
+     * Checks XeTwitterLink equals method.
+     * @throws Exception If some problem inside
+     */
+    @Test
+    public void equalsAndHashCodeEqualTest() throws Exception {
```
Comment (ggajos) [inline src/test/java/org/takes/facets/auth/social/XeTwitterLinkTest.java]: @ikhvostenkov is not following http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names convention

File: src/test/java/org/takes/facets/auth/social/XeTwitterLinkTest.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                        new XeTwitterLink("testToken")
+                    )
+                ).body()
+            ),
+            XhtmlMatchers.hasXPaths(
+                "/root/links/link[@rel='takes:twitter']"
+            )
+        );
+    }
+    /**
```
Comment (ggajos) [inline src/test/java/org/takes/facets/auth/social/XeTwitterLinkTest.java]: @ikhvostenkov let's add space before comment

File: src/main/java/org/takes/facets/auth/social/XeTwitterLink.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+     * Ctor.
+     * @param token Twitter token
+     */
+    public XeTwitterLink(final CharSequence token) {
+        super(XeTwitterLink.make("takes:twitter", token));
+    }
+
+    /**
+     * Generate link.
+     * @param rel Related
```
Comment (ggajos) [inline src/main/java/org/takes/facets/auth/social/XeTwitterLink.java]: @ikhvostenkov let's explain what `Related` is

**Comment by ggajos** [comment]
@ikhvostenkov few comments

**Comment by ikhvostenkov** [comment]
@ggajos Thank you for your comments. All issues above have been fixed. Please, review.

File: src/test/java/org/takes/facets/auth/social/XeTwitterLinkTest.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            )
+        );
+    }
+
+    /**
+     * XeTwitterLink can compare objects.
+     * @throws Exception If some problem inside
+     */
+    @Test
+    public void compareObjectsTest() throws Exception {
```
Comment (ggajos) [inline src/test/java/org/takes/facets/auth/social/XeTwitterLinkTest.java]: @ikhvostenkov let's remove Test suffix

**Comment by ggajos** [comment]
@ikhvostenkov last minor comment

**Comment by ikhvostenkov** [comment]
@ggajos Fixed. Check this, please.

**Comment by ggajos** [comment]
@ikhvostenkov thanks

**Comment by ggajos** [comment]
@rultor merge please

**Comment by yegor256** [comment]
@ikhvostenkov hm... I don't think it will work. what is "token"? Twitter gives me application ID and secret key. There is no such thing as token. Where did you get this information? Add links to the source code (into javadoc). See also `PsTwitter`

**Comment by ikhvostenkov** [comment]
@yegor256 Maybe I am mistaking here, but as I understood from Twitter API:

Token
The oauth_token parameter typically represents a user’s permission to share access to their account with your application. There are a few authentication requests where this value is not passed or is a different form of token, but those are covered in detail in Obtaining access tokens. For most general-purpose requests, you will use what is referred to as an access token. You can generate a valid access token for your account on the settings page for your application at dev.twitter.com/apps.
oauth_token	370773112-GmHxMAgYyLbNEtIKZeRNFsMKPR9EyMZeS9weJAEb

[Read more](https://dev.twitter.com/oauth/overview/authorizing-requests)

And then you are using this oauth_token in the [authorize request](https://dev.twitter.com/oauth/reference/get/oauth/authorize).

**Comment by ikhvostenkov** [comment]
@yegor256 Any updates here?

**Comment by yegor256** [comment]
@ikhvostenkov try to create your own app in Twitter and try to get that "token" from it. You will see this message in the console:

> This access token can be used to make API requests on your own account's behalf. Do not share your access token secret with anyone.

Would be strange to put that token to the URL.

**Comment by ikhvostenkov** [comment]
@yegor256 You are right, this is not that token, but this token requested before providing user with OAuth page with ```http://twitter.com/oauth/request_token``` request. After That user is redirected to OAuth page with ```https://api.twitter.com/oauth/authorize``` and with the token from previous request. If the goal is to create link for requesting request_token, I can redo, but as I understand goal is to create link to OAuth page. Also try to open any web site and sign in with twitter you will get  ```https://api.twitter.com/oauth/authorize``` request with oauth_token nether with app_id no with secret_key. [Here is information from Twitter API] (https://dev.twitter.com/web/sign-in/implementing)

**Comment by ggajos** [comment]
@ikhvostenkov any progress here?

**Comment by ggajos** [comment]
@yegor256 @ikhvostenkov this PR should be closed?

**Comment by ikhvostenkov** [comment]
@ggajos I need confirmation from @yegor256 site if this is correct implementation or not.

File: src/main/java/org/takes/facets/auth/social/XeTwitterLink.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+     * Generate link.
+     * @param rel Link relation type
+     * @param token Twitter token
+     * @return Source
+     */
+    private static XeSource make(final CharSequence rel,
+        final CharSequence token) {
+        return new XeLink(
+            rel,
+            new Href("https://api.twitter.com/oauth/authorize")
```
Comment (ekondrashev) [inline src/main/java/org/takes/facets/auth/social/XeTwitterLink.java]: Shouldn't it be `https://api.twitter.com/oauth/authenticate` instead as per the [tutorial](https://dev.twitter.com/web/sign-in/implementing) step2 ?

**Comment by ekondrashev** [comment]
@yegor256 perhaps you want target `XeSource` to perform [post](https://dev.twitter.com/oauth/reference/post/oauth/request_token) request in order to obtain `oauth_token`? I'm not sure it is good idea, but if your opinion is different then there should be a possibility to get returned `auth_token_secret` and `oauth_callback_confirmed`. Pretty much for one class. Perhaps there should be a puzzle created for mentioned post request logic encapsulation?

**Comment by yegor256** [comment]
@ikhvostenkov looks like the entire design is wrong here, if we really need to get the token first. We should not create this `XeTwitterLink` class, but instead implement the functionality of token retrieval inside `PsTwitter`. see the problem?

**Comment by ikhvostenkov** [comment]
@yegor256 I got your point. This makes sense. I see 2 possibilities here: change this class to get Xembly for oauth_token link or just cancel this PR and close original ticket. I need your architect and original issue starter decision here.

**Comment by yegor256** [comment]
@ikhvostenkov looks like we have to close the original issue and re-design the entire Twitter OAuth mechanism, because it is different from Facebook/Github/Google+

**Comment by ikhvostenkov** [comment]
@davvd @yegor256 @ggajos PR closed due to errors in current Twitter OAuth mechanism.

**Comment by davvd** [comment]
@elenavolokhova please, review this ticket for compliance with our [QA rules](http://www.teamed.io/qa.html)

**Comment by elenavolokhova** [comment]
@ggajos 
According to our [quality rules](http://www.teamed.io/qa.html):
 > Comments were mostly about design problems, not cosmetic issues.
 
 There were mostly cosmetic issues found during review. Try to pay attention to bigger design problems and the overall quality of the code.
Please confirm that this new rule is clear.

**Comment by ggajos** [comment]
@elenavolokhova thanks, I will

**Comment by elenavolokhova** [comment]
@davvd Quality is acceptable

**Comment by davvd** [comment]
> @davvd Quality is acceptable

@elenavolokhova thank you, let's try to make it "good" next time

**Comment by davvd** [comment]
@ggajos I added 10 mins to @elenavolokhova (for QA review) in transaction 59063322; I just added **22 mins** to your account, many thanks for your contribution (AP-9JT14172JE753994L).. 504 hours and 36 mins spent here; review comments (c=7) added as a bonus; added +22 to your rating, now it is equal to [+1389](http://www.netbout.com/b/35094?open=rating)

**Pull request closed without merging**